File: apps/server/src/application/pilot/commands/create-pilot-product.command.ts
// src/application/pilot/commands/create-pilot-product.command.ts

import { Data } from 'effect'
import * as S from 'effect/Schema'
import {
  CreatePilotProductRequest,
  type CreatePilotProductRequest as CreatePilotProductRequestType,
} from '@maison-amane/api'

import { CorrelationIdSchema, UserIdSchema } from '../../../domain/shared'

// ============================================
// UNVALIDATED PRODUCT DATA (from API boundary)
// ============================================

export type UnvalidatedProductData = CreatePilotProductRequestType

// ============================================
// CREATE PILOT PRODUCT COMMAND
// ============================================

const PilotProductCreationCommandSchema = S.TaggedStruct('CreatePilotProductCommand', {
  data: CreatePilotProductRequest,
  correlationId: CorrelationIdSchema,
  userId: UserIdSchema,
  timestamp: S.Date,
})

export type PilotProductCreationCommand = typeof PilotProductCreationCommandSchema.Type

export const makePilotProductCreationCommand = (
  params: Omit<PilotProductCreationCommand, '_tag'>
): PilotProductCreationCommand =>
  Data.case<PilotProductCreationCommand>()({ _tag: 'CreatePilotProductCommand', ...params })

File: apps/server-java/src/main/java/com/maisonamane/application/pilot/command/CreatePilotProductCommand.java
package com.maisonamane.application.pilot.command;

import com.maisonamane.domain.shared.CorrelationId;
import com.maisonamane.domain.shared.UserId;
import java.time.Instant;
import java.util.Objects;

/**
 * Command to create a new Pilot Product.
 * Maps Effect-TS TaggedStruct("CreatePilotProductCommand") to Java record.
 */
public record CreatePilotProductCommand(
    UnvalidatedProductData data,
    CorrelationId correlationId,
    UserId userId,
    Instant timestamp
) {

    public CreatePilotProductCommand {
        Objects.requireNonNull(data, "Data cannot be null");
        Objects.requireNonNull(correlationId, "CorrelationId cannot be null");
        Objects.requireNonNull(userId, "UserId cannot be null");
        Objects.requireNonNull(timestamp, "Timestamp cannot be null");
    }

    public static CreatePilotProductCommand of(
        UnvalidatedProductData data,
        CorrelationId correlationId,
        UserId userId,
        Instant timestamp
    ) {
        return new CreatePilotProductCommand(data, correlationId, userId, timestamp);
    }
}
